### PR TITLE
Vastly improve performance

### LINKED ISSUE #30: Normalize paths to `/`?
This is useful for checking filename patterns without needing to handle different separators.
The [`path-slash` crate](https://crates.io/crates/path-slash) might help.

Possibly relates to #28

### LINKED ISSUE #31: Use `.gitattributes` from rev?
Right now, gengo uses the `.gitattributes` file from the filesystem. E.g. changes to `.gitattributes` that haven't been committed can affect the file attributes (vendored, detectable, etc.).

github-linguist uses the `.gitattributes` *of the analyzed rev.* Until you commit your changes to `.gitattributes`, it will not affect github-linguist.

Right now I don't know which is preferable.

### LINKED ISSUE #154: Drop `git2` for `gix`
As mentioned in https://github.com/o2sh/onefetch/pull/1152#issuecomment-1703915072, `gix` will likely greatly improve performance.

### LINKED ISSUE #155: Parallelize
As mentioned in https://github.com/o2sh/onefetch/pull/1152#issuecomment-1703871936, this has potential for for performance improvements.

cc @Byron

### PR DESCRIPTION
**Pull request by Byron**
The approach taken here it so provide multiple steps of the transformation which should be reviewed commit by commit. The goal is to start out sticking with the original structure and then transform it into the fastest possible solution. The latter will look quite different, thus the step-wise approach.

### Proposed Review Workflow

I'd appreciate if you could make the changes as you see fit and push directly into this PR. `gh pr checkout 157` and `git push` should do the trick for that. That way we can spare each other slow round-trips that typically aren't necessary as you can more quickly do what's needed.

### Review Notes

- Slashes are universally used now, let's see what happens with the tests on windows.
- `test/javascript` branch now has to be adjusted to contain `.gitattributes` as I changed the `.gitattributes` sources. Feel free to just drop the commit though if it's too uncertain.

### Tasks

* [x] `git2` to `gix` - 5.5x faster than baseline (yes, really 😁)
* [x] use index instead of manual tree traversal (enables parallelization, later submodule handling) 
    - 5.7x faster than baseline, webkit finishes in 15s now (down from 2min), an ~8x speedup (we are used to this now ;)). Pretty sure the extreme slowdown is to the inefficient handling of attributes in `git2` - the exposed API is really hard to get fast as well.
* [x] parallelize `git`-based ingestion - on a small repo, the speedup is *just* ~20x; on WebKit it now takes 1.9s on a hot cache, that's 60x faster; on Linux it's 40x faster, on Rust it's 55x faster (without submodules)
* [x] submodule support - all timings are the same, but for Rust it changes to only 8.5x faster as it now considers a lot more files. 
* [x] Use attributes from `rev` only - this makes runs on `Rust` or `Webkit` about 100-150ms faster. Not massive, but not nothing to sneeze at either.

### Correctness Fixes

* don't assume all paths are UTF8 encoded, but skip those who don't transform correctly (typically on windows)

### Affected Issues

* Fixes #154
* Fixes #155
* Probably fixes #31 
* Fixes #30 

<details><summary>Performance Data</summary>

##### `git2` to `gix`G

```
gitoxide ( improvements) [$]
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengo
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):      68.9 ms ±   0.6 ms    [User: 61.0 ms, System: 7.2 ms]
  Range (min … max):    67.7 ms …  71.4 ms    43 runs

Benchmark 2: gengo
  Time (mean ± σ):     380.8 ms ±   1.3 ms    [User: 146.6 ms, System: 232.4 ms]
  Range (min … max):   378.6 ms … 383.1 ms    10 runs

Summary
  /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo ran
    5.53 ± 0.05 times faster than gengo
```

##### Use index, single-threaded

```
gitoxide ( improvements) [$] took 7s
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengo
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):      66.8 ms ±   0.9 ms    [User: 59.6 ms, System: 6.6 ms]
  Range (min … max):    65.5 ms …  69.3 ms    44 runs

Benchmark 2: gengo
  Time (mean ± σ):     381.4 ms ±   1.5 ms    [User: 147.2 ms, System: 232.3 ms]
  Range (min … max):   377.8 ms … 383.0 ms    10 runs

Summary
  /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo ran
    5.71 ± 0.08 times faster than gengo
```

#### Use index, multi-threaded

```
gitoxide ( improvements) [$] took 3s
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengo
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):      19.2 ms ±   0.4 ms    [User: 74.0 ms, System: 23.2 ms]
  Range (min … max):    18.5 ms …  20.8 ms    144 runs

Benchmark 2: gengo
  Time (mean ± σ):     382.2 ms ±   2.6 ms    [User: 147.0 ms, System: 233.4 ms]
  Range (min … max):   378.5 ms … 385.8 ms    10 runs

Summary
  /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo ran
   19.92 ± 0.43 times faster than gengo
```

```
linux ( master) +369 -819 [!] took 8s
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengos
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):     655.8 ms ±   5.3 ms    [User: 4941.1 ms, System: 482.2 ms]
  Range (min … max):   644.6 ms … 662.3 ms    10 runs

Benchmark 2: gengos
Error: Failed to run command 'gengos': No such file or directory (os error 2)

linux ( master) +369 -819 [!] took 8s
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengo
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):     669.3 ms ±  15.9 ms    [User: 4918.0 ms, System: 485.3 ms]
  Range (min … max):   653.7 ms … 695.1 ms    10 runs

Benchmark 2: gengo
  Time (mean ± σ):     27.069 s ±  0.308 s    [User: 11.664 s, System: 15.397 s]
  Range (min … max):   26.679 s … 27.570 s    10 runs

Summary
  /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo ran
   40.44 ± 1.06 times faster than gengo
```

#### Rust-repo without submodules

```
rust ( master) via 🐍
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengo
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):     144.1 ms ±   6.7 ms    [User: 920.6 ms, System: 137.7 ms]
  Range (min … max):   139.2 ms … 170.6 ms    21 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 2: gengo
  Time (mean ± σ):      8.052 s ±  0.145 s    [User: 2.771 s, System: 5.225 s]
  Range (min … max):    7.849 s …  8.297 s    10 runs

Summary
  /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo ran
   55.86 ± 2.80 times faster than gengo
```

#### Rust-repo - with submodules

Note the `gengo` doesn't consider submodules, so has far less work to do.

```
rust ( master) +11 -11 via 🐍 took 7s
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengo
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):     932.6 ms ±   3.2 ms    [User: 6813.0 ms, System: 831.0 ms]
  Range (min … max):   929.2 ms … 939.4 ms    10 runs

Benchmark 2: gengo
  Time (mean ± σ):      7.967 s ±  0.100 s    [User: 2.749 s, System: 5.205 s]
  Range (min … max):    7.812 s …  8.100 s    10 runs

Summary
  /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo ran
    8.54 ± 0.11 times faster than gengo
```

#### Webkit profile

This shows that gathering an index for the repository takes a lot of time and it can't be parallelized. The cost of doing business.

Screenshot 2023-09-03 at 15 00 33

#### Linux - 1% performance loss due to 10 attributes, instead of just 5

```
❯ hyperfine -N --warmup 1 /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo gengo
Benchmark 1: /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
  Time (mean ± σ):     651.3 ms ±   5.2 ms    [User: 4939.9 ms, System: 384.2 ms]
  Range (min … max):   643.2 ms … 658.2 ms    10 runs

Benchmark 2: gengo
  Time (mean ± σ):     644.8 ms ±   4.4 ms    [User: 4899.4 ms, System: 372.5 ms]
  Range (min … max):   637.5 ms … 650.2 ms    10 runs

Summary
  gengo ran
    1.01 ± 0.01 times faster than /Users/byron/dev/github.com/spenserblack/gengo/target/release/gengo
```

</details>

### CODE REVIEW

**Comment by spenserblack** [comment]
This is awesome! I'll review more when I get a chance, but just some quick responses: 

> I recommend either keeping support for both gengo- and linguist- attributes, but would prefer to only keep the linguist-* set for 1% more

I'd like to *not* support `linguist-` attributes. While this crate takes a lot of inspiration from linguist, I want the freedom to diverge from how linguist does things. For example, using different names for languages. And right now the `-detectable` behavior is (intentionally) different.

******

Note to self: submodules should be detected as vendored by default.

**Comment by Byron** [comment]
Great, take your time!

I will fix CI and undo the change to attributes. Also I plan to make a new release of `gix` with a QoL improvement, but besides that what you see should be pretty much 'it'.

----

When looking at linguist and remembering that it took 5min on a repo, I couldn't stop to wonder if they run this at GitHub scale. If so, how much time must be wasted 🥺.

**Comment by Byron** [comment]
CI should be fixed now, and the `gix` version used here will stay the same after all. Thus it should definitely be ready now :).

**Comment by spenserblack** [comment]
> When looking at linguist and remembering that it took 5min on a repo, I couldn't stop to wonder if they run this at GitHub scale. If so, how much time must be wasted 🥺.

Performance is probably much better on GitHub's hardware than my own limited VM :laughing: but yeah, it does raise that question. And Linguist has had some issues -- recently, one of its heuristics was a ReDOS that stopped some files from rendering or getting highlighted. Ironically, that ReDOS was introduced in an attempt to get `heuristics.yml` to use more generic regexes for a Rust port of linguist :laughing:

There's also the question of what exactly causes the time spent for the Linux repo. Is it simply the file count, or are there file(s) that severely hit their Bayesian classifier, or something else? Usage as a library might be more performant as a binary, too. If the classifier is set up for each time you run the binary, but only once for the library, for example.

******

I guess it's time to admit my hopeful thinking: I know that GitHub switched their code search to a Rust implementation for more performance. Could there perhaps be a Rust project for GitHub's language detection? :thinking: :laughing:

**Comment by spenserblack** [comment]
BTW don't worry about the failing tests for now. I'll update the `.gitattributes` in `test/javascript` and update the snapshots as needed.

**Comment by Byron** [comment]
> I guess it's time to admit my hopeful thinking: I know that GitHub switched their code search to a Rust implementation for more performance. Could there perhaps be a Rust project for GitHub's language detection? 🤔 😆

I actually had a feeling :D! But then you said there are differences and incompatibilities which makes that less likely to happen, so they would probably rather do their own RiR. Time will tell what's going to happen though :).

Do you have an idea why people would add `linguist-*` attributes in the first place? Does that affect GitHub language statistics?

**Comment by spenserblack** [comment]
> Do you have an idea why people would add `linguist-*` attributes in the first place?

One of the most common is probably `linguist-language`. Linguist does a good job of detecting languages, but it's not perfect. Personally, I've used `.vscode/*.json linguiist-language=JSON-with-comments` a lot, since linguist often detects "JSON with comments" as "JSON *without* comments."

People use `linguist-documentation`, `linguist-vendored`, and `linguist-generated` almost interchangeably as a way to tell linguist to ignore certain files. `linguist-generated` is supposed to also exclude files from PR diffs by default (see how `Cargo.lock` diffs aren't rendered in PRs), which some like and some don't.

File: gengo/src/analysis/mod.rs [hunk 1]
```diff
@@ -42,21 +65,8 @@ impl Analysis {
...
-    }
-}
-
-impl<'map> IntoIterator for &'map Analysis {
-    type Item = (&'map PathBuf, &'map Entry);
-    type IntoIter = Iter<'map>;
-
-    fn into_iter(self) -> Self::IntoIter {
-        self.iter()
+impl std::fmt::Debug for Analysis {
```
Comment (spenserblack) [inline gengo/src/analysis/mod.rs:68]: Nitpick: Why prefix with `std::fmt::`? `Debug` is one of the default imports AFAIK.
```suggestion
impl Debug for Analysis {
```

File: gengo/src/analysis/mod.rs [hunk 2]
```diff
@@ -42,21 +65,8 @@ impl Analysis {
...
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
```
Comment (spenserblack) [inline gengo/src/analysis/mod.rs:70]: Looks like this `Debug` implementation removed the type's identity. Do you think that it could be added back, so the output is
```
Analysis {
...
```

File: gengo/src/builder.rs [hunk 3]
```diff
@@ -50,22 +48,21 @@ impl<P: AsRef<Path>> Builder<P> {
...
-        let repository = match Repository::discover(self.repository_path) {
+        let repository = match gix::discover(self.repository_path) {
             Ok(r) => r,
-            Err(e) => match e.code() {
-                ErrorCode::NotFound => {
-                    return Err(Box::new(Error::with_source(ErrorKind::NoRepository, e)));
-                }
-                _ => return Err(Box::new(e)),
-            },
+            Err(gix::discover::Error::Discover(err)) => {
```
Comment (spenserblack) [inline gengo/src/builder.rs:53]: Styling nitpick: import `gix::discover::Error` so this full path doesn't have to be used.

File: gengo/src/lib.rs [hunk 4]
```diff
@@ -39,127 +42,243 @@ const GLOB_MATCH_OPTIONS: MatchOptions = MatchOptions {
...
+    fn from_index(
+        root: BString,
+        index: gix::index::State,
+    ) -> (Self, Vec<(BString, gix::ObjectId)>) {
+        let (entries, path_storage) = index.into_entries();
+        let mut submodules = Vec::new();
+        let entries: Vec<_> = entries
+            .into_iter()
+            .filter(|e| {
+                if e.mode == gix::index::entry::Mode::COMMIT {
```
Comment (spenserblack) [inline gengo/src/lib.rs:110]: Nitpick: Import `gix::index::entry::Mode` so this doesn't need to be so verbose.
Maybe make this a `match` since it's `e.mode == X` repeated.

File: gengo/src/lib.rs [hunk 5]
```diff
@@ -39,127 +42,243 @@ const GLOB_MATCH_OPTIONS: MatchOptions = MatchOptions {
...
         Ok(())
     }
 
-    fn analyze_blob<P: AsRef<Path>>(
+    fn analyze_blob(
         &self,
-        filepath: P,
-        blob: &Blob,
-        results: &mut IndexMap<PathBuf, Entry>,
+        filepath: impl AsRef<Path>,
```
Comment (spenserblack) [inline gengo/src/lib.rs:206]: Just so I can learn: why switch from a generic type to `impl AsRef<Path>`?

File: gengo/src/lib.rs [hunk 6]
```diff
@@ -39,127 +42,243 @@ const GLOB_MATCH_OPTIONS: MatchOptions = MatchOptions {
...
-        let generated = self
-            .get_boolean_attr(&filepath, "gengo-generated")?
-            .unwrap_or_else(|| self.is_generated(&filepath, contents));
-        let documentation = self
-            .get_boolean_attr(&filepath, "gengo-documentation")?
-            .unwrap_or_else(|| self.is_documentation(&filepath, contents));
-        let vendored = self
-            .get_boolean_attr(&filepath, "gengo-vendored")?
-            .unwrap_or_else(|| self.is_vendored(&filepath, contents));
+        let generated = attrs[1]
```
Comment (spenserblack) [inline gengo/src/lib.rs:248]: Just a note since I found this bit a little hard to follow at first.
```suggestion
        // NOTE Unspecified attributes are filtered out, so if .is_set() is false, .is_unset() is true.
        let generated = attrs[1]
```

**Review by spenserblack — COMMENTED** [review]
Thanks again!

File: gengo-bin/Cargo.toml [hunk 7]
```diff
@@ -19,6 +19,7 @@ color = ["owo-colors", "gengo/owo-colors"]
 
 [dependencies]
 clap = { version = "4", features = ["derive", "wrap_help"] }
+gix-for-configuration-only = { package = "gix", version = "0.52.0", default-features = false, features = ["max-performance"] }
```
[the remark this answers is not in the document]
Comment (Byron) [reply gengo-bin/Cargo.toml:22]: This makes sure that the application uses the fastest possible configuration, whereas the library ideally uses only the features that it needs. From that point of view, even the current `max-performance-safe` might be more than necessary.

[the remark this answers is not in the document]
Comment (spenserblack) [reply gengo-bin/Cargo.toml:22]: Would it make sense for gengo to have
```toml
[features]
max-performance = ["gix/max-performance"]
```

And then `gengo-bin` could have

```toml
gengo = { ..., features = ["max-performance"] }
```

I feel like it's a bit less hacky to re-export features instead of adding a dependency that doesn't actually get used at the level it's imported.

File: gengo/src/analysis/mod.rs (hunk 1 above)
Comment (Byron) [reply gengo/src/analysis/mod.rs:68]: There is no reason in particular. It seems you are working on this branch right now (just pushed an update for windows), and I welcome that you make changes as you see fit.

File: gengo-bin/Cargo.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (Byron) [reply gengo-bin/Cargo.toml:22]: I thought about this, but believe it's not an option to keep forward feature toggles. Instead, the officially endorsed way is to only setup the `gix` features that are actually required, and leave all others for applications to configure.

Maybe one day `cargo` allows to set features of crates in the dependency tree without having them as direct dependency. It's absolutely up to you though.

File: gengo/src/analysis/mod.rs (hunk 2 above)
Comment (Byron) [reply gengo/src/analysis/mod.rs:70]: Actually I tried but it's not possible without implementing parts of the Debug impl by hand. Since it's used for debugging only, I don't think it's a problem.
One could, of course, manually write `f.write_str("Analyis {")?;` and wrap it up with `f.write_str("}")`, but formatting will look different unless a lot of care is taken that wouldn't make sense for `insta` I think.

File: gengo/src/lib.rs (hunk 5 above)
Comment (Byron) [reply gengo/src/lib.rs:206]: It's less noisy to my eyes and a syntax that was added to provide a reasonable alternative to the one it (partially) replaces. I'd consider it more idiomatic, too.

File: gengo/src/lib.rs (hunk 6 above)
Comment (Byron) [reply gengo/src/lib.rs:250]: This might be a hint that the API for how to deal with attribute values can be improved. I was also trying hard to assure it's actually doing the same as `git2` or, even better, `git` would do. I *think* it's correct, and if you have ideas on how to make it more intuitive, please let me know.

**Comment by Byron** [comment]
Thanks for the review.
Could you do me a small favor and apply the style changes (and any other changes) yourself? It's not worth your time to write about it as it's so easy to fix at least while in front of a computer. Besides that, I am happy to answer any question that might come up, or address bigger issues that are more than a quick-fix. Thank again.

**Comment by spenserblack** [comment]
> Could you do me a small favor and apply the style changes (and any other changes) yourself?

Yup! I wasn't that clear, but some of these are to dos for myself. I'd appreciate if quickly skimmed over my style changes/refactoring later, though, just to confirm the changes are purely code style, not affecting behavior.

**Comment by spenserblack** [comment]
:nauseated_face: Per-OS snapshot tests certainly have their downsides :laughing: I should've learned from repofetch.

******

@Byron as the gix expert, do you have any recommendations on detecting files from submodules as vendored? My lazy self was thinking of just adding a `is_submodule` boolean to the stack :laughing:

It would probably be cleanest to implement this in the `Vendored` struct, but that would probably complicate constructing the struct.

**Comment by spenserblack** [comment]
:thinking: Now that I think about it, while traversing submodules is *really cool,* I wonder if it should be dropped. I'd want to mark submodules as vendored, but that is *very specific to git,* and could block, or at least make much more difficult to implement, other file sources. And submodules are technically other repos, and usually not code that belongs to the main project...

**Comment by Byron** [comment]
> @Byron as the gix expert, do you have any recommendations on detecting files from submodules as vendored? My lazy self was thinking of just adding a `is_submodule` boolean to the stack 😆

I thought the same - with that flag it should be possible to implement any logic. And yes, it's `false` just for the first repo :D.



> 🤔 Now that I think about it, while traversing submodules is _really cool,_ I wonder if it should be dropped. I'd want to mark submodules as vendored, but that is _very specific to git,_ and could block, or at least make much more difficult to implement, other file sources. And submodules are technically other repos, and usually not code that belongs to the main project...

Not traversing submodules is faster, of course, and maybe for that reason alone it could be a flag to be configured. However, I think it's nothing to skip as, at least I, am interesting how much of the project is vendored. Maybe that's entirely the wrong thing to do though, and it's better to look at how `linguist` does it.

**Comment by Byron** [comment]
I thought about this again and definitely think that submodules should not be ignored. However, it's the question what to do with them by default or… how to allow users to affect what they are considered. There is no problem in assigning attributes to submodules, and I think this could be used to affect if they should be vendored or not. This reminds me, maybe one should check for `is_unset()` to disable the vendor-by-default behaviour?

**Comment by spenserblack** [comment]
How would `.gitattributes` get resolved for a submodule's files, by the way? I've never experimented with it, but I would assume that they wouldn't be able to "affect" a submodule's files. If you're modifying submodule `/A`, then the following `.gitattributes` in the root repo wouldn't really do anything, right?

```gitattributes
# /.gitattributes
A/* eol=lf
```

Anyway, for now I'm really excited about this PR, so I'd be happy to merge this and continue the discussion on how to handle submodules later. So perhaps just an `is_submodule` flag on the stack for now?

**Comment by spenserblack** [comment]
OK, thinking about it a bit more, it might be nice for `Vendored` to have an `add_vendored_dir` method, or something similar. Any submodule we encounter could have its path registered to that. The code would be a bit cleaner and more testable that way IMO.

**Comment by Byron** [comment]
> How would `.gitattributes` get resolved for a submodule's files, by the way? I've never experimented with it, but I would assume that they wouldn't be able to "affect" a submodule's files. If you're modifying submodule `/A`, then the following `.gitattributes` in the root repo wouldn't really do anything, right?

Actually it would - it's perfectly viable to use complete paths of submodule files (as relative to the superproject root) for an attribute query in the superproject. But that's not what I meant - I merely meant to check the path of the `submodule/worktree` against the superproject to get attributes for configuration. With that, whatever default you chose could be overridden.


> Anyway, for now I'm really excited about this PR, so I'd be happy to merge this and continue the discussion on how to handle submodules later. So perhaps just an `is_submodule` flag on the stack for now?

That would be enough to declare all submodule files vendored, I think.

> OK, thinking about it a bit more, it might be nice for `Vendored` to have an `add_vendored_dir` method, or something similar. Any submodule we encounter could have its path registered to that. The code would be a bit cleaner and more testable that way IMO.

Maybe that doesn't matter (yet) for merging the PR like you mentioned. However, I don't think submodules should *always* be vendored, there should be en escape hatch, that just treats them like ordinary code that is part of the tree (like it's done now). Then tying submodules into vendor-handling too much seems counter-intuitive. But again, in the end it doesn't matter as long as the code does what it should and you define what that is.

**Comment by spenserblack** [comment]
Going to merge as-is, we can handle submodules in a separate PR. I'll see if I can get something set up.

I agree that submodules shouldn't be *always* vendored, but I do think that should be the default, since "vendored" basically means "files from another project that are provided in this repo."

**Comment by spenserblack** [comment]
@all-contributors add @Byron for code and userTesting
